mitmproxy/proxy/protocol2/test/test_websocket.py

```diff
@@ -0,0 +1,104 @@
+from unittest import mock
+
+from . import tutils
+from .. import commands
+from .. import events
+from .. import websocket
+from mitmproxy.test import tflow
+
+
+def test_simple(tctx):
+    f = tutils.Placeholder()
+    tctx.server.connected = True
+    playbook = tutils.playbook(websocket.WebsocketLayer(tctx, tflow.twebsocketflow().handshake_flow))
+
+    with mock.patch("os.urandom") as m:
+        m.return_value = b"\x10\x11\x12\x13"
+        assert (
+            playbook
+            << commands.Hook("websocket_start", f)
+            >> events.HookReply(-1, None)
+            >> events.DataReceived(tctx.client, b"\x82\x85\x10\x11\x12\x13Xt~\x7f\x7f")  # Frame with payload b"Hello"
+            << commands.Hook("websocket_message", f)
+            >> events.HookReply(-1, None)
+            << commands.SendData(tctx.server, b"\x82\x85\x10\x11\x12\x13Xt~\x7f\x7f")
+            >> events.DataReceived(tctx.server, b'\x81\x05Hello')  # Frame with payload "Hello"
+            << commands.Hook("websocket_message", f)
+            >> events.HookReply(-1, None)
+            << commands.SendData(tctx.client, b'\x81\x05Hello')
+            >> events.DataReceived(tctx.client, b'\x88\x82\x10\x11\x12\x13\x13\xf9')
+            << commands.SendData(tctx.server, b'\x88\x82\x10\x11\x12\x13\x13\xf9')
+            << commands.SendData(tctx.client, b'\x88\x02\x03\xe8')
+            << commands.Hook("websocket_end", f)
+            >> events.HookReply(-1, None)
+            << commands.CloseConnection(tctx.server)
```

Comment:
Ok, it looks like I haven't made the current closing semantics clear yet. It's a bit nonobvious: If a connection is completed, all we want to do is closing the client connection. Once the proxy server (the thing that does IO) notices that the client connection is shut down, it will automatically close all other connections as well. 

If we close a server connection, the client connection will remain open. This has the purpose as e.g. a HTTP/2 connection could have multiple server connections at the same time because a user changed the hostname in a script for one of the requests.

With the current code here, we rely on the client actually closing the connection, otherwise we end up with a stale connection, as both ends (client and proxy) wait for the other end to close. Does that make sense and do you see the problem?